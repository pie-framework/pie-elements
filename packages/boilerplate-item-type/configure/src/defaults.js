// Should be exactly the same as controller/defaults.js
export default {
  model: {
    prompt: 'Question Prompt goes here',
    promptEnabled: true
  },
  configuration: {
    prompt: {
      label: 'Prompt',
      settings: true,
    }
  },
};
