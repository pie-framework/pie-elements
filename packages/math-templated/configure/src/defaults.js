// Should be exactly the same as controller/defaults.js
export default {
  model: {
    markup: '',
    prompt: '',
    promptEnabled: true,
    teacherInstructionsEnabled: true,
    rationale: 'Rationale goes here',
    rationaleEnabled: true,
    spellCheckEnabled: true,
    playerSpellCheckEnabled: true,
    responses: {},
    element: 'math-templated',
    equationEditor: '8',
  },
  configuration: {
    baseInputConfiguration: {
      html: { disabled: true },
      audio: { disabled: false },
      video: { disabled: false },
      image: { disabled: false },
    },
    prompt: {
      label: 'Prompt',
      settings: true,
      inputConfiguration: {
        audio: { disabled: false },
        video: { disabled: false },
        image: { disabled: false },
      },
      required: false,
    },
    spellCheck: {
      label: 'Spellcheck',
      settings: false,
      enabled: true,
    },
    editSource: {
      label: 'Edit Source',
      settings: false,
      enabled: false,
    },
    playerSpellCheck: {
      label: 'Student Spellcheck',
      settings: true,
      enabled: true,
    },
    teacherInstructions: {
      settings: true,
      label: 'Teacher Instructions',
      inputConfiguration: {
        audio: { disabled: false },
        video: { disabled: false },
        image: { disabled: false },
      },
      required: false,
    },
    rationale: {
      settings: true,
      label: 'Rationale',
      inputConfiguration: {
        audio: { disabled: false },
        video: { disabled: false },
        image: { disabled: false },
      },
      required: false,
    },
    template: {
      inputConfiguration: {
        audio: { disabled: false },
        video: { disabled: false },
        image: { disabled: false },
      },
    },
    maxImageWidth: {
      teacherInstructions: 300,
      prompt: 300,
      rationale: 300,
    },
    maxImageHeight: {
      teacherInstructions: 300,
      prompt: 300,
      rationale: 300,
    },
    mathMlOptions: {
      mmlOutput: false,
      mmlEditing: false,
    },
    language: {
      settings: false,
      label: 'Specify Language',
      enabled: false,
    },
    languageChoices: {
      label: 'Language Choices',
      options: [],
    },
  },
};
