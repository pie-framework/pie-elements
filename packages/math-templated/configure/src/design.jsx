import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import cloneDeep from 'lodash/cloneDeep';
import pick from 'lodash/pick';
import throttle from 'lodash/throttle';
import {
  InputContainer,
  settings,
  layout,
} from '@pie-lib/pie-toolbox/config-ui';
import {
  ALL_PLUGINS,
  EditableHtml,
} from '@pie-lib/pie-toolbox/editable-html';
import { dropdown } from '@pie-lib/pie-toolbox/code/config-ui/settings';

import Response from './response';
import { processMarkup } from './markupUtils';

const { Panel, toggle } = settings;

const createElementFromHTML = (htmlString) => {
  const div = document.createElement('div');
  div.innerHTML = htmlString.trim();

  return div;
};

class Design extends React.Component {
  static propTypes = {
    model: PropTypes.object.isRequired,
    configuration: PropTypes.object.isRequired,
    classes: PropTypes.object.isRequired,
    onModelChanged: PropTypes.func.isRequired,
    onConfigurationChanged: PropTypes.func.isRequired,
    imageSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired,
    }),
    uploadSoundSupport: PropTypes.object,
  };

  state = {};

  componentDidMount() {
    const { model: { slateMarkup } } = this.props;

    this.setState({ markup: slateMarkup });

  }

  handleChange = (key, value) => {
    const { onModelChanged, model } = this.props;
    const updatedModel = cloneDeep(model);
    updatedModel[key] = value;
    onModelChanged(updatedModel);
  };

  onResponseChange = (response, index) => {
    const { model, onModelChanged } = this.props;
    const newModel = { ...model };

    newModel.responses[index] = response;
    onModelChanged(newModel);
  };

  onChangeResponse = (index, newVal) => {
    const { model, onModelChanged } = this.props;
    const { responses } = model;

    if (!responses[index]) {
      responses[index] = [{ answer: newVal || '', id: 'response' + index, allowSpaces: true }];
    } else {
      responses[index][0].answer = newVal || '';
    }

    onModelChanged({ ...model, responses });
  };

  onResponsesChanged = (responses) => {
    this.props.onModelChanged({ ...this.props.model, responses });
  };

  onChange = (markup) => {
    const {
      model: { responses },
      onModelChanged,
    } = this.props;
    const newResponses = {};
    const domMarkup = createElementFromHTML(markup);
    const responseAreas = domMarkup.querySelectorAll('[data-type="math_templated"]');

    responseAreas.forEach((element) => {
      const { value, index: dataIndex } = element.dataset;

      if (!value) {
        element.dataset.value = `R ${dataIndex}`;
      }

      newResponses[dataIndex] = responses[dataIndex] || {
        // TODO: Initialize with customizable defaults
        allowSpaces: true,
        validation: 'symbolic',
        allowTrailingZeros: false,
        ignoreOrder: false,
        answer: '',
        alternates: {},
      };

      element.dataset.index = dataIndex;
    });

    const processedMarkup = processMarkup(markup);

    const callback = () =>
      onModelChanged({
        ...this.props.model,
        slateMarkup: domMarkup.innerHTML,
        responses: newResponses,
        markup: processedMarkup
      });

    this.setState({ cachedChoices: undefined }, callback);
  };

  onHandleAreaChange = throttle(
    (nodes) => {
      const {
        model: { responses },
        onModelChanged,
      } = this.props;
      const { cachedResponses } = this.state;

      if (!nodes) {
        return;
      }

      const newChoices = responses ? cloneDeep(responses) : {};
      const newCachedResponses = cachedResponses ? cloneDeep(cachedResponses) : {};

      nodes.forEach((node) => {
        const keyForNode = node.data.get('index');

        if (!newChoices[keyForNode] && newCachedResponses[keyForNode]) {
          Object.assign(newChoices, pick(newCachedResponses, keyForNode));

          if (newCachedResponses.hasOwnProperty(keyForNode)) {
            delete newCachedResponses[keyForNode];
          }
        } else {
          Object.assign(newCachedResponses, pick(newChoices, keyForNode));

          if (newChoices.hasOwnProperty(keyForNode)) {
            delete newChoices[keyForNode];
          }
        }
      });

      const callback = () => onModelChanged({ ...this.props.model, responses: newChoices });

      this.setState({ cachedResponses: newCachedResponses }, callback);
    },
    500,
    { trailing: false, leading: true },
  );

  onBlur = (e) => {
    const { relatedTarget, currentTarget } = e || {};

    function getParentWithRoleTooltip(element, depth = 0) {
      // only run this max 16 times
      if (!element || depth >= 16) return null;

      const parent = element.offsetParent;

      if (!parent) return null;

      if (parent.getAttribute('role') === 'tooltip') {
        return parent;
      }

      return getParentWithRoleTooltip(parent, depth + 1);
    }

    function getDeepChildDataKeypad(element, depth = 0) {
      // only run this max 4 times
      if (!element || depth >= 4) return null;

      const child = element?.children?.[0];

      if (!child) return null;

      if (child.attributes && child.attributes['data-keypad']) {
        return child;
      }

      return getDeepChildDataKeypad(child, depth + 1);
    }

    const parentWithTooltipRole = getParentWithRoleTooltip(relatedTarget);
    const childWithDataKeypad = parentWithTooltipRole ? getDeepChildDataKeypad(parentWithTooltipRole) : null;

    if (!relatedTarget || !currentTarget || !childWithDataKeypad?.attributes['data-keypad']) {
      this.setState({ activeAnswerBlock: '' });
    }
  };

  render() {
    const {
      classes,
      configuration,
      imageSupport,
      model,
      onConfigurationChanged,
      onModelChanged,
      uploadSoundSupport,
    } = this.props;

    const {
      baseInputConfiguration = {},
      contentDimensions = {},
      prompt = {},
      rationale = {},
      settingsPanelDisabled,
      teacherInstructions = {},
      language = {},
      languageChoices = {},
      spellCheck = {},
      playerSpellCheck = {},
      maxImageWidth = {},
      maxImageHeight = {},
      mathMlOptions = {},
      template = {},
      editSource = {},
      ignoreOrder: cIgnoreOrder = {},
      allowTrailingZeros: cAllowTrailingZeros = {},
      partialScoring = {},
      maxResponseAreas,
    } = configuration || {};

    const {
      errors,
      promptEnabled,
      rationaleEnabled,
      spellCheckEnabled,
      teacherInstructionsEnabled,
      toolbarEditorPosition,
      responses,
      equationEditor,
    } = model || {};

    const {
      choices: choicesErrors = {},
      prompt: promptError,
      rationale: rationaleError,
      responseAreas: responseAreasError,
      teacherInstructions: teacherInstructionsError,
      responses: responsesErrors,
    } = errors || {};

    const panelSettings = {
      'language.enabled': language.settings && toggle(language.label, true),
      language: language.settings && language.enabled && dropdown(languageChoices.label, languageChoices.options),
    };

    const panelProperties = {
      teacherInstructionsEnabled: teacherInstructions.settings && toggle(teacherInstructions.label),
      rationaleEnabled: rationale.settings && toggle(rationale.label),
      promptEnabled: prompt.settings && toggle(prompt.label),
      spellCheckEnabled: spellCheck.settings && toggle(spellCheck.label),
      playerSpellCheckEnabled: playerSpellCheck.settings && toggle(playerSpellCheck.label),
      'editSource.enabled': editSource?.settings && toggle(editSource.label, true),
      partialScoring: partialScoring.settings && toggle(partialScoring.label),
    };

    const defaultImageMaxWidth = maxImageWidth && maxImageWidth.prompt;
    const defaultImageMaxHeight = maxImageHeight && maxImageHeight.prompt;

    const toolbarOpts = {
      position: toolbarEditorPosition === 'top' ? 'top' : 'bottom',
    };

    const getPluginProps = (props = {}, baseInputConfiguration = {}) => ({
      ...baseInputConfiguration,
      ...props,
    });

    return (
      <layout.ConfigLayout
        dimensions={contentDimensions}
        hideSettings={settingsPanelDisabled}
        settings={
          <Panel
            model={model}
            configuration={configuration}
            onChangeModel={(updatedModel) => onModelChanged(updatedModel)}
            onChangeConfiguration={onConfigurationChanged}
            groups={{
              Settings: panelSettings,
              Properties: panelProperties,
            }}
          />
        }
      >
        {teacherInstructionsEnabled && (
          <InputContainer
            label={teacherInstructions.label}
            className={classes.promptHolder}
          >
            <EditableHtml
              className={classes.prompt}
              markup={model.teacherInstructions || ''}
              onChange={(value) => this.handleChange('teacherInstructions', value)}
              imageSupport={imageSupport}
              nonEmpty={false}
              error={teacherInstructionsError}
              toolbarOpts={toolbarOpts}
              pluginProps={getPluginProps(
                teacherInstructions?.inputConfiguration,
                baseInputConfiguration
              )}
              spellCheck={spellCheckEnabled}
              maxImageWidth={
                (maxImageWidth && maxImageWidth.teacherInstructions) ||
                defaultImageMaxWidth
              }
              maxImageHeight={
                (maxImageHeight && maxImageHeight.teacherInstructions) ||
                defaultImageMaxHeight
              }
              uploadSoundSupport={uploadSoundSupport}
              languageCharactersProps={[
                { language: 'spanish' },
                { language: 'special' },
              ]}
              mathMlOptions={mathMlOptions}
            />
            {teacherInstructionsError && (
              <div className={classes.errorText}>{teacherInstructionsError}</div>
            )}
          </InputContainer>
        )}

        {promptEnabled && (
          <InputContainer
            label={prompt.label}
            className={classes.promptHolder}
          >
            <EditableHtml
              className={classes.prompt}
              markup={model.prompt}
              onChange={(value) => this.handleChange('prompt', value)}
              imageSupport={imageSupport}
              nonEmpty={false}
              disableUnderline
              error={promptError}
              toolbarOpts={toolbarOpts}
              pluginProps={getPluginProps(
                teacherInstructions?.inputConfiguration,
                baseInputConfiguration
              )}
              spellCheck={spellCheckEnabled}
              maxImageWidth={defaultImageMaxWidth}
              maxImageHeight={defaultImageMaxHeight}
              uploadSoundSupport={uploadSoundSupport}
              languageCharactersProps={[
                { language: 'spanish' },
                { language: 'special' },
              ]}
              mathMlOptions={mathMlOptions}
            />
            {promptError && (
              <div className={classes.errorText}>{promptError}</div>
            )}
          </InputContainer>
        )}

        <Typography className={classes.title}>Response Template</Typography>

        <EditableHtml
          activePlugins={ALL_PLUGINS}
          toolbarOpts={{ position: 'top' }}
          spellCheck={spellCheckEnabled}
          pluginProps={getPluginProps(
            template?.inputConfiguration,
            baseInputConfiguration
          )}
          responseAreaProps={{
            type: 'math-templated',
            respAreaToolbar: null,
            error: () => choicesErrors,
            onHandleAreaChange: this.onHandleAreaChange,
            maxResponseAreas: maxResponseAreas,
          }}
          className={classes.markup}
          markup={model.slateMarkup}
          onChange={this.onChange}
          imageSupport={imageSupport}
          disableImageAlignmentButtons={true}
          onBlur={this.onBlur}
          disabled={false}
          highlightShape={false}
          error={responseAreasError}
          uploadSoundSupport={uploadSoundSupport}
          languageCharactersProps={[
            { language: 'spanish' },
            { language: 'special' },
          ]}
          mathMlOptions={mathMlOptions}
        />

        <Typography className={classes.title}>Define Response</Typography>

        <InputContainer
          label="Response Template Equation Editor"
          className={classes.selectContainer}
        >
          <Select
            className={classes.select}
            onChange={(event) =>
              this.handleChange('equationEditor', event.target.value)
            }
            value={equationEditor}
          >
            <MenuItem value="non-negative-integers">
              Numeric - Non-Negative Integers
            </MenuItem>
            <MenuItem value="integers">Numeric - Integers</MenuItem>
            <MenuItem value="decimals">Numeric - Decimals</MenuItem>
            <MenuItem value="fractions">Numeric - Fractions</MenuItem>
            <MenuItem value={1}>Grade 1 - 2</MenuItem>
            <MenuItem value={3}>Grade 3 - 5</MenuItem>
            <MenuItem value={6}>Grade 6 - 7</MenuItem>
            <MenuItem value={8}>Grade 8 - HS</MenuItem>
            <MenuItem value={'geometry'}>Geometry</MenuItem>
            <MenuItem value={'advanced-algebra'}>
              Advanced Algebra
            </MenuItem>
            <MenuItem value={'statistics'}>Statistics</MenuItem>
            <MenuItem value={'item-authoring'}>Item Authoring</MenuItem>
          </Select>
        </InputContainer>

        {Object.keys(responses || {}).map((responseKey, idx) => {
          const response = responses[responseKey];

          return (
            <Response
              key={responseKey}
              responseKey={responseKey}
              mode={equationEditor}
              response={response}
              onResponseChange={this.onResponseChange}
              index={idx}
              cIgnoreOrder={cIgnoreOrder}
              cAllowTrailingZeros={cAllowTrailingZeros}
              error={responsesErrors && responsesErrors[idx]}
            />
          );
        })}

        {rationaleEnabled && (
          <InputContainer
            label={rationale.label}
            className={classes.promptHolder}
          >
            <EditableHtml
              className={classes.prompt}
              markup={model.rationale || ''}
              onChange={(value) => this.handleChange('rationale', value)}
              imageSupport={imageSupport}
              nonEmpty={false}
              error={rationaleError}
              toolbarOpts={toolbarOpts}
              pluginProps={getPluginProps(rationale?.inputConfiguration, {
                ...baseInputConfiguration,
                math: {
                  controlledKeypadMode: false,
                },
              })}
              spellCheck={spellCheckEnabled}
              maxImageWidth={
                (maxImageWidth && maxImageWidth.rationale) ||
                defaultImageMaxWidth
              }
              maxImageHeight={
                (maxImageHeight && maxImageHeight.rationale) ||
                defaultImageMaxHeight
              }
              uploadSoundSupport={uploadSoundSupport}
              languageCharactersProps={[
                { language: 'spanish' },
                { language: 'special' },
              ]}
              mathMlOptions={mathMlOptions}
            />
            {rationaleError && (
              <div className={classes.errorText}>{rationaleError}</div>
            )}
          </InputContainer>
        )}
      </layout.ConfigLayout>
    );
  }
}

export default withStyles((theme) => ({
  promptHolder: {
    width: '100%',
    paddingTop: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 2,
  },
  prompt: {
    width: '100%',
  },
  errorText: {
    color: theme.palette.error.main,
    fontSize: '0.75rem',
    marginTop: theme.spacing.unit,
  },
  markup: {
    width: '100%',
    marginTop: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 3,
  },
  selectContainer: {
    width: '100%',
    marginTop: theme.spacing.unit * 2,
  },
  select: {
    width: '100%',
  },
  title: {
    fontSize: theme.typography.fontSize * 1.25,
    fontWeight: 'bold',
  }
}))(Design);
