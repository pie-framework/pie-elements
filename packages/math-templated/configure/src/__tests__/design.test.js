import React from 'react';
import { shallow } from 'enzyme';
import { Design } from '../design';
import defaults from '../defaults';
import { InputContainer } from '@pie-lib/pie-toolbox/config-ui';
import { EditableHtml } from '@pie-lib/pie-toolbox/editable-html';
import Select from '@material-ui/core/Select';
import Tooltip from '@material-ui/core/Tooltip';

jest.mock('@pie-lib/pie-toolbox/config-ui', () => ({
    layout: {
        ConfigLayout: (props) => <div>{props.children}</div>,
    },
    settings: {
        Panel: (props) => <div onChange={props.onChange} />,
        toggle: jest.fn(),
        dropdown: jest.fn(),
    },
    InputContainer: (props) => <div {...props}>{props.children}</div>,
    EditableHtml: (props) => <div {...props}>{props.children}</div>,
}));

describe('Render Main Component', () => {
    let wrapper, onChange;
    let model = defaults.model;
    let configuration = defaults.configuration;

    beforeEach(() => {
        onChange = jest.fn();

        wrapper = shallow(
            <Design
                classes={{}}
                model={model}
                configuration={configuration}
                onModelChanged={onChange}
                imageSupport={{}}
                uploadSoundSupport={{}}
            />,
        );
        
    });

    it('Match Snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    describe('UI Rendering', () => {
        it('renders teacher instructions input when enabled', () => {
            wrapper.setProps({
                model: { ...model, teacherInstructionsEnabled: true },
            });
            expect(wrapper.find(InputContainer).at(0).prop('label')).toEqual('Teacher Instructions');
        });

        it('renders prompt input when enabled', () => {
            wrapper.setProps({
                model: { ...model, promptEnabled: true },
            });
            expect(wrapper.find(InputContainer).at(1).prop('label')).toEqual('Prompt');
        });

        it('renders rationale input when enabled', () => {
            wrapper.setProps({
                model: { ...model, rationaleEnabled: true },
            });
            expect(wrapper.find(InputContainer).at(3).prop('label')).toEqual('Rationale');
        });

        it('renders response template and markup correctly', () => {
            const mockMarkup = '<div data-type="math_templated"></div>';
            wrapper.setState({ markup: mockMarkup });
            wrapper.setProps({ model: { ...model, slateMarkup: mockMarkup } });

            const editableHtml = wrapper.find(EditableHtml);
            expect(editableHtml.exists()).toBe(true);
            expect(editableHtml.at(2).prop('markup')).toEqual(mockMarkup);
        });

        it('renders equation editor select correctly', () => {
            const inputContainer = wrapper.find(InputContainer).findWhere(node => node.prop('label') === 'Response Template Equation Editor');
            expect(inputContainer.exists()).toBe(true);

            const select = inputContainer.find(Select);
            expect(select.exists()).toBe(true);
        });

        it('renders tooltip with correct title', () => {
            const tooltip = wrapper.find(Tooltip);
            expect(tooltip.exists()).toBe(true);
            expect(tooltip.prop('title')).toEqual(
                'Validation requirements:\nCorrect answers should not be blank.\nEach answer defined for a response area should be unique.\nThere should be at least 1 and at most 10 response areas defined.'
            );
        });
    });

    describe('logic', () => {
        it('handles onChange event correctly for teacher instructions', () => {
            wrapper.setProps({ model: { ...model, teacherInstructionsEnabled: true } });

            const inputContainer = wrapper.find(InputContainer).at(0);
            inputContainer.find(EditableHtml).simulate('change', '<p>New teacher instructions</p>');

            expect(onChange).toHaveBeenCalledWith({ ...model, teacherInstructions: '<p>New teacher instructions</p>' });
        });

        it('handles onChange event correctly for prompt', () => {
            wrapper.setProps({ model: { ...model, promptEnabled: true } });

            const inputContainer = wrapper.find(InputContainer).at(1);
            inputContainer.find(EditableHtml).simulate('change', '<p>New prompt</p>');

            expect(onChange).toHaveBeenCalledWith({ ...model, prompt: '<p>New prompt</p>' });
        });

        it('handles onChange event correctly for rationale', () => {
            wrapper.setProps({ model: { ...model, rationaleEnabled: true } });

            const inputContainer = wrapper.find(InputContainer).at(3);
            inputContainer.find(EditableHtml).simulate('change', '<p>New rationale</p>');

            expect(onChange).toHaveBeenCalledWith({ ...model, rationale: '<p>New rationale</p>' });
        });

        it('not renders prompt input when disabled', () => {
            wrapper.setProps({ model: { ...model, promptEnabled: false } });

            expect(wrapper.find(InputContainer).at(1).prop('label')).not.toEqual('Prompt');
        });

        it('updates state correctly when markup changes', () => {
            const mockMarkup = '<p>New markup content</p>';
            wrapper.setState({ markup: mockMarkup });
            wrapper.setProps({ model: { ...model, slateMarkup: mockMarkup } });

            wrapper.find(EditableHtml).at(2).simulate('change', mockMarkup);

            expect(wrapper.state('markup')).toEqual(mockMarkup);
        });

        it('updates model correctly when equation editor value changes', () => {
            const newValue = 'integers';
            const inputContainer = wrapper.find(InputContainer).findWhere(node => node.prop('label') === 'Response Template Equation Editor');
            const select = inputContainer.find(Select);

            select.simulate('change', { target: { value: newValue } });

            expect(onChange).toHaveBeenCalledWith({ ...model, equationEditor: newValue });
        });
    });

});