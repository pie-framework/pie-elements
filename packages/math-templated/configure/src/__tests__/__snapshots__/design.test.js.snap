// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Main Component Match Snapshot 1`] = `
<ConfigLayout
  dimensions={Object {}}
  settings={
    <Panel
      configuration={
        Object {
          "allowTrailingZeros": Object {
            "enabled": true,
            "label": "Allow Trailing Zeros",
            "settings": false,
          },
          "baseInputConfiguration": Object {
            "audio": Object {
              "disabled": false,
            },
            "blockquote": Object {
              "disabled": true,
            },
            "h3": Object {
              "disabled": true,
            },
            "html": Object {
              "disabled": true,
            },
            "image": Object {
              "disabled": false,
            },
            "textAlign": Object {
              "disabled": true,
            },
            "video": Object {
              "disabled": false,
            },
          },
          "editSource": Object {
            "enabled": false,
            "label": "Edit Source",
            "settings": false,
          },
          "ignoreOrder": Object {
            "enabled": true,
            "label": "Ignore Order",
            "settings": false,
          },
          "language": Object {
            "enabled": false,
            "label": "Specify Language",
            "settings": false,
          },
          "languageChoices": Object {
            "label": "Language Choices",
            "options": Array [],
          },
          "mathMlOptions": Object {
            "mmlEditing": false,
            "mmlOutput": false,
          },
          "maxImageHeight": Object {
            "prompt": 300,
            "rationale": 300,
            "teacherInstructions": 300,
          },
          "maxImageWidth": Object {
            "prompt": 300,
            "rationale": 300,
            "teacherInstructions": 300,
          },
          "maxResponseAreas": 10,
          "partialScoring": Object {
            "label": "Allow Partial Scoring",
            "settings": true,
          },
          "playerSpellCheck": Object {
            "enabled": true,
            "label": "Student Spellcheck",
            "settings": true,
          },
          "prompt": Object {
            "inputConfiguration": Object {
              "audio": Object {
                "disabled": false,
              },
              "image": Object {
                "disabled": false,
              },
              "video": Object {
                "disabled": false,
              },
            },
            "label": "Prompt",
            "required": false,
            "settings": true,
          },
          "rationale": Object {
            "inputConfiguration": Object {
              "audio": Object {
                "disabled": false,
              },
              "image": Object {
                "disabled": false,
              },
              "video": Object {
                "disabled": false,
              },
            },
            "label": "Rationale",
            "required": false,
            "settings": true,
          },
          "spellCheck": Object {
            "enabled": true,
            "label": "Spellcheck",
            "settings": false,
          },
          "teacherInstructions": Object {
            "inputConfiguration": Object {
              "audio": Object {
                "disabled": false,
              },
              "image": Object {
                "disabled": false,
              },
              "video": Object {
                "disabled": false,
              },
            },
            "label": "Teacher Instructions",
            "required": false,
            "settings": true,
          },
          "template": Object {
            "inputConfiguration": Object {
              "audio": Object {
                "disabled": false,
              },
              "image": Object {
                "disabled": false,
              },
              "video": Object {
                "disabled": false,
              },
            },
          },
        }
      }
      groups={
        Object {
          "Properties": Object {
            "editSource.enabled": false,
            "partialScoring": undefined,
            "playerSpellCheckEnabled": undefined,
            "promptEnabled": undefined,
            "rationaleEnabled": undefined,
            "spellCheckEnabled": false,
            "teacherInstructionsEnabled": undefined,
          },
          "Settings": Object {
            "language": false,
            "language.enabled": false,
          },
        }
      }
      model={
        Object {
          "allowTrailingZerosDefault": false,
          "equationEditor": "8",
          "ignoreOrderDefault": false,
          "markup": "",
          "playerSpellCheckEnabled": true,
          "prompt": "",
          "promptEnabled": true,
          "rationale": "",
          "rationaleEnabled": true,
          "responses": Object {},
          "spellCheckEnabled": true,
          "teacherInstructions": "",
          "teacherInstructionsEnabled": true,
          "toolbarEditorPosition": "bottom",
          "validationDefault": "literal",
        }
      }
      onChangeModel={[Function]}
    />
  }
>
  <InputContainer
    label="Teacher Instructions"
  >
    <EditableHtml
      allowValidation={false}
      imageSupport={Object {}}
      languageCharactersProps={
        Array [
          Object {
            "language": "spanish",
          },
          Object {
            "language": "special",
          },
        ]
      }
      markup=""
      mathMlOptions={
        Object {
          "mmlEditing": false,
          "mmlOutput": false,
        }
      }
      maxImageHeight={300}
      maxImageWidth={300}
      nonEmpty={false}
      onChange={[Function]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": false,
          },
          "blockquote": Object {
            "disabled": true,
          },
          "h3": Object {
            "disabled": true,
          },
          "html": Object {
            "disabled": true,
          },
          "image": Object {
            "disabled": false,
          },
          "textAlign": Object {
            "disabled": true,
          },
          "video": Object {
            "disabled": false,
          },
        }
      }
      spellCheck={true}
      toolbarOpts={
        Object {
          "position": "bottom",
        }
      }
      uploadSoundSupport={Object {}}
    />
  </InputContainer>
  <InputContainer
    label="Prompt"
  >
    <EditableHtml
      allowValidation={false}
      disableUnderline={true}
      imageSupport={Object {}}
      languageCharactersProps={
        Array [
          Object {
            "language": "spanish",
          },
          Object {
            "language": "special",
          },
        ]
      }
      markup=""
      mathMlOptions={
        Object {
          "mmlEditing": false,
          "mmlOutput": false,
        }
      }
      maxImageHeight={300}
      maxImageWidth={300}
      nonEmpty={false}
      onChange={[Function]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": false,
          },
          "blockquote": Object {
            "disabled": true,
          },
          "h3": Object {
            "disabled": true,
          },
          "html": Object {
            "disabled": true,
          },
          "image": Object {
            "disabled": false,
          },
          "textAlign": Object {
            "disabled": true,
          },
          "video": Object {
            "disabled": false,
          },
        }
      }
      spellCheck={true}
      toolbarOpts={
        Object {
          "position": "bottom",
        }
      }
      uploadSoundSupport={Object {}}
    />
  </InputContainer>
  <div>
    <WithStyles(Typography)
      component="div"
    >
      Response Template
    </WithStyles(Typography)>
    <WithStyles(Tooltip)
      classes={
        Object {
          "tooltip": undefined,
        }
      }
      disableFocusListener={true}
      disableTouchListener={true}
      placement="right"
      title="Validation requirements:
Correct answers should not be blank.
Each answer defined for a response area should be unique.
There should be at least 1 and at most 10 response areas defined."
    >
      <pure(InfoIcon)
        color="primary"
        fontSize="small"
      />
    </WithStyles(Tooltip)>
  </div>
  <EditableHtml
    activePlugins={
      Array [
        "bold",
        "html",
        "extraCSSRules",
        "italic",
        "underline",
        "strikethrough",
        "bulleted-list",
        "numbered-list",
        "image",
        "math",
        "languageCharacters",
        "blockquote",
        "h3",
        "table",
        "video",
        "audio",
        "responseArea",
        "redo",
        "undo",
        "superscript",
        "subscript",
      ]
    }
    allowValidation={false}
    disableImageAlignmentButtons={true}
    disabled={false}
    highlightShape={false}
    imageSupport={Object {}}
    languageCharactersProps={
      Array [
        Object {
          "language": "spanish",
        },
        Object {
          "language": "special",
        },
      ]
    }
    mathMlOptions={
      Object {
        "mmlEditing": false,
        "mmlOutput": false,
      }
    }
    onBlur={[Function]}
    onChange={[Function]}
    onDone={[Function]}
    pluginProps={
      Object {
        "audio": Object {
          "disabled": false,
        },
        "blockquote": Object {
          "disabled": true,
        },
        "h3": Object {
          "disabled": true,
        },
        "html": Object {
          "disabled": true,
        },
        "image": Object {
          "disabled": false,
        },
        "textAlign": Object {
          "disabled": true,
        },
        "video": Object {
          "disabled": false,
        },
      }
    }
    responseAreaProps={
      Object {
        "error": [Function],
        "maxResponseAreas": 10,
        "onHandleAreaChange": [Function],
        "respAreaToolbar": null,
        "type": "math-templated",
      }
    }
    spellCheck={true}
    toolbarOpts={
      Object {
        "position": "top",
      }
    }
    uploadSoundSupport={Object {}}
  />
  <WithStyles(Typography)>
    Define Response
  </WithStyles(Typography)>
  <InputContainer
    label="Response Template Equation Editor"
  >
    <WithStyles(WithFormControlContext(Select))
      onChange={[Function]}
      value="8"
    >
      <WithStyles(MenuItem)
        value="non-negative-integers"
      >
        Numeric - Non-Negative Integers
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="integers"
      >
        Numeric - Integers
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="decimals"
      >
        Numeric - Decimals
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="fractions"
      >
        Numeric - Fractions
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={1}
      >
        Grade 1 - 2
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={3}
      >
        Grade 3 - 5
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={6}
      >
        Grade 6 - 7
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value={8}
      >
        Grade 8 - HS
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="geometry"
      >
        Geometry
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="advanced-algebra"
      >
        Advanced Algebra
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="statistics"
      >
        Statistics
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        value="item-authoring"
      >
        Item Authoring
      </WithStyles(MenuItem)>
    </WithStyles(WithFormControlContext(Select))>
  </InputContainer>
  <InputContainer
    label="Rationale"
  >
    <EditableHtml
      allowValidation={false}
      imageSupport={Object {}}
      languageCharactersProps={
        Array [
          Object {
            "language": "spanish",
          },
          Object {
            "language": "special",
          },
        ]
      }
      markup=""
      mathMlOptions={
        Object {
          "mmlEditing": false,
          "mmlOutput": false,
        }
      }
      maxImageHeight={300}
      maxImageWidth={300}
      nonEmpty={false}
      onChange={[Function]}
      onDone={[Function]}
      pluginProps={
        Object {
          "audio": Object {
            "disabled": false,
          },
          "blockquote": Object {
            "disabled": true,
          },
          "h3": Object {
            "disabled": true,
          },
          "html": Object {
            "disabled": true,
          },
          "image": Object {
            "disabled": false,
          },
          "math": Object {
            "controlledKeypadMode": false,
          },
          "textAlign": Object {
            "disabled": true,
          },
          "video": Object {
            "disabled": false,
          },
        }
      }
      spellCheck={true}
      toolbarOpts={
        Object {
          "position": "bottom",
        }
      }
      uploadSoundSupport={Object {}}
    />
  </InputContainer>
</ConfigLayout>
`;
